apm-agent-plugins/apm-servlet-plugin/src/main/java/co/elastic/apm/agent/servlet/AsyncInstrumentation.java

```diff
@@ -54,87 +54,130 @@
-    @VisibleForAdvice
-    public static class StartAsyncAdvice {
-
-        @Advice.OnMethodExit(suppress = Throwable.class)
-        private static void onExitStartAsync(@Advice.This HttpServletRequest request, @Advice.Return AsyncContext asyncContext) {
-            if (tracer != null) {
-                final Transaction transaction = tracer.currentTransaction();
-                if (transaction != null) {
-                    request.setAttribute(TRANSACTION_ATTRIBUTE, transaction);
-                    if (asyncHelper != null) {
-                        asyncHelper.getForClassLoaderOfClass(AsyncContext.class).onExitStartAsync(asyncContext);
+        @Override
+        public Class<?> getAdviceClass() {
+            return StartAsyncAdvice.class;
+        }
+
+        @VisibleForAdvice
+        public static class StartAsyncAdvice {
+
+            @Advice.OnMethodExit
+            private static void onExitStartAsync(@Advice.This HttpServletRequest request, @Advice.Return AsyncContext asyncContext) {
+                if (tracer != null) {
+                    final Transaction transaction = tracer.currentTransaction();
+                    if (transaction != null) {
+                        request.setAttribute(TRANSACTION_ATTRIBUTE, transaction);
+                        if (asyncHelper != null) {
+                            asyncHelper.getForClassLoaderOfClass(AsyncContext.class).onExitStartAsync(asyncContext);
+                        }
                     }
                 }
             }
         }
     }
 
+    public static class AsyncContextInstrumentation extends AsyncInstrumentation {
+        @Override
+        public ElementMatcher<? super NamedElement> getTypeMatcherPreFilter() {
+            return nameContains("AsyncContext");
+        }
+
+        @Override
+        public ElementMatcher<? super TypeDescription> getTypeMatcher() {
+            return not(isInterface())
+                .and(hasSuperType(named("javax.servlet.AsyncContext")));
+        }
+
+        @Override
+        public ElementMatcher<? super MethodDescription> getMethodMatcher() {
+            return isPublic()
+                .and(named("start"))
+                .and(takesArguments(Runnable.class));
+        }
+
+        @Override
+        public Class<?> getAdviceClass() {
+            return AsyncContextStartAdvice.class;
+        }
+
+        @VisibleForAdvice
+        public static class AsyncContextStartAdvice {
+
+            @Advice.OnMethodEnter(suppress = Throwable.class)
+            private static void onEnterAsyncContextStart(@Advice.Argument(value = 0, readOnly = false) @Nullable Runnable runnable) {
+                if (tracer != null) {
+                    final Transaction transaction = tracer.currentTransaction();
+                    if (transaction != null && runnable != null && asyncHelper != null) {
+                        runnable = asyncHelper.getForClassLoaderOfClass(AsyncContext.class).wrapRunnable(runnable, transaction);
```

Comment:
wouldn't it be easier to just call the `wrapRunnable` method on the `tracer` instance? Then you could also remove the `wrapRunnable` from the helper.

Reply:
When implementing I imagined it will do some additional stuff, but probably no need for that as it is now